rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,175 @@
+        string          signature format
+        byte[]          signature blob
+
+Unlike the standard `sign@openssh.com` request, `sign@goteleport.com` expects to receive digested data (pre-hashed) rather than the raw message to digest on the agent side. This makes it possible to perform signatures with algorithms other than the ssh signing algorithms (e.g. `ssh-rsa`, `ssh-rsa-cert-v01@openssh.com`). For example, this enables the use of `RSASSA-PSS` algorithms used in TLS 1.3 handshakes.
+
+Hash name should be the stringified representation of a golang `crypto.Hash` value. For example, `SHA-1`, `SHA-256`, or `SHA-512`.
+
+Salt length can be empty for algorithms that don't use a salt, or a positive integer for those that do (such as `RSASSA-PSS`). Salt length can also be set to `auto` to automatically use the largest salt length possible during signing, which can be auto-detected during verification.
+
+**Prompt MFA Challenge Extension:**
+
+This extension can be used to issue an MFA challenge prompt to the user's local machine, which enables support for MFA functionality including `tsh` MFA login and per-session MFA verification.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          add-mfa-key@goteleport.com
+        string          proxy address
+        []byte          challenge blob (json)
+
+Where challenge blob is a json encoded `api/client/proto.MFAAuthenticateChallenge`.
+
+The resulting challenge response can then be used for MFA verification.
+
+        byte            SSH_AGENT_SUCCESS
+        []byte          challenge response blob (json)
+
+Where challenge response blob is a json encoded `api/client/proto.MFAAuthenticateResponse`.
+
+Note: The protobuf structs used above are subject to change, which may lead to backwards compatibility concerns. In this case, this extension should be taken into consideration before making changes to these structs.
+
+### Security
+
+#### SSH Agent Forwarding Risks
+
+SSH Agent forwarding introduces an inherent risk. When a user does `ssh -A` or `tsh ssh -A`, their forwarded keys could be used by a user on the remote machine with OS permissions equal to or above the remote user. Still, these forwarded keys can only be used as long as the user maintains the agent forwarding sessions, since agent forwarding does not allow keys to be exported (only certificates). This security principle constrains some potential options, such as providing the user's raw private key via the `list-keys@goteleport.com` extension.
+
+However, even with the new extensions constrained in this way, a user can abuse the forwarded agent to reissue certificates with new private keys held on the remote host. This raises a security concern that is not present in standard ssh agent forwarding, since the user's login session can essentially be exported to the remote host via a reissue command, rather than being contingent on the forwarding agent session providing access to a private key on the local host.
```

Comment:
could you please add more info on it

Reply:
I added another sentence to clarify that this can be done with the `GenerateUserCerts` rpc. Please let me know if/what you'd like me to clarify further.

Reply:
thank you, it looks good to me